sui_programmability/examples/games/sources/frenemies.move

```diff
@@ -0,0 +1,387 @@
+    const EAlreadyUpdatedScorecardThisEpoch: u64 = 0;
+
+    fun init(ctx: &mut TxContext) {
+        let epoch = tx_context::epoch(ctx);
+        transfer::share_object(
+            Leaderboard {
+                id: object::new(ctx),
+                prev_epoch_stakes: table::new(ctx),
+                epoch,
+                start_epoch: epoch,
+            }
+        )
+    }
+
+    /// Get the current assignment for `addr`.
+    /// Always returns the same assignment for a fixed address A and epoch E.
+    fun get_assignment(addr: address, state: &SuiSystemState): Assignment {
+        let validators = validator_set::active_validators(sui_system::validators(state));
+        let len = vector::length(validators);
+
+        // derive assignment from the stake of top validator + address.
+        // this means that for a fixed address A and epoch E, the assignment is deterministic
+        let top_stake = (validator::total_stake(vector::borrow(validators, 0)) as u256);
+        let addr = address::to_u256(addr);
+        let assignment_seed = (addr ^ top_stake);
+        let validator_idx = assignment_seed % (len as u256);
+        let validator = validator::sui_address(vector::borrow(validators, (validator_idx as u64)));
+        let goal = ((assignment_seed % 3) as u8);
+        Assignment {
+            validator,
+            goal,
+        }
+    }
+
+    /// Register the transaction sender for the frenemies game by sending them a 0 scorecard with
+    /// an initial assignment
+    public entry fun register(state: &SuiSystemState, ctx: &mut TxContext) {
+        let sender = tx_context::sender(ctx);
+        transfer::transfer(
+            Scorecard {
+                id: object::new(ctx),
+                assignment: get_assignment(sender, state),
+                score: 0,
+                participation: 0,
+                epoch: tx_context::epoch(ctx),
+            },
+            sender
+        )
+    }
+
+    /// Update `scorecard` with the results from the last epoch.
+    /// Note: relies on someone calling this at least once/epoch to preserve
+    /// the record of validator scores
+    public entry fun update(
+        scorecard: &mut Scorecard, state: &SuiSystemState, leaderboard: &mut Leaderboard, ctx: &mut TxContext
+    ) {
+        let cur_epoch = tx_context::epoch(ctx);
+        if (cur_epoch > leaderboard.epoch) {
+            // if first call in new epoch, update leaderboard with stake rankings for the previous epoch.
+            // Note: there's a pragmatically safe, but theoretically unsafe assumption here that
+            // `cur_epoch == leaderboard.epoch + 1` because at least one player will update their scorecard
+            // each epoch. if this assumption is violated via 0 calls to `update` during epoch `N`, players
+            // will not get credit for scores during epoch `N - 1`.
+            update_epoch_info(state, leaderboard, cur_epoch - 1)
```

Comment:
I think it makes more sense to use `cur_epoch` here and maybe `cur_epoch + 1` for the `scorecard`'s epoch below. We should use `cur_epoch` here because the stake distribution stored in the system state is the stake distribution that's used in our system throughout the current epoch so using the current epoch number would be more intuitive.